README.md
# Danmakanvas

## What is Danmakanvas?

Danmakanvas is a lightweight danmaku engine built in JavaScript (for HTML5 Canvas) that can be embedded into webpages to provide visual examples of danmaku patterns. It provides functionality for creating basic bullets, as well as simulated tasks, and is primarily developed for the sake of allowing cross-platform (Except *cough* INTERNET EXPLORER *cough*) Danmaku prototyping and visualization.

Danmakanvas was first written May-June 2016 (v1) and was, to be frank, OK at best, with a plethora of SetInterval problems and limited usage. However, in January 2018, this old project was slightly rewritten and added to this website (v2) in a form that was usable... if there was only one canvas per page. In April-May 2018, Danmakanvas was rewritten to allow for multiple canvases on the same webpage.

This repository serves as a home for the repository for development on v2.2 onwards so that it is not forcibly coupled to sparen.github.io's repo and also to provide documentation on usage.

## Limitations

Due to the nature of this project having shifted to pattern prototyping, there is currently no support for custom PNG graphics, player scripts, multiple attacks in a row, sound, etc. 

This project is currently not accepting contributions in the form of code. If bugs are encountered, please post an Issue here on GitHub.

## Components

To run Danmakanvas, you must have an HTML source file. This source file must contain a canvas with an ID, height, and width. The HTML source must include two scripts - the engine, and what I call a 'PluralModule'. Further details will be provided below.

### The Game Engine

The game engine contains two control functions - `createNewGame(canvasid, title)` and `stopGame(canvasid, clearFrame)`. The HTML file must call `createNewGame()` in order to start or restart a Danmakanvas Instance. Stopping the canvas via `stopGame()` will leave the canvas showing the last rendered frame unless the optional parameter `clearFrame` is set to true. The canvas ID passed as a parameter is used as the primary identifier of the Danmakanvas Instance. 

Each new game keeps track of the bullets, current plurals running, the canvas, and the frame counter, among other things. The NewGame objects have functionality to start a game associated with a canvas ID. As of v2.3, they are no longer capable of resetting themselves internally and must be reset from outside (which is what `stopGame()` does). The actual code to run in the instances is located in the Plural Modules, which will be discussed shortly.

When a new game is started, it sets an interval to run a main update loop for the provided canvas ID. This ensures that multiple Danmakufu Instances can run in parallel. The main update loop (`update_main()`) wipes the canvas (to allow the updates to be drawn), then directs the plural controller to update (this will be discussed soon). It checks every bullet and forces them to update, then removes bullets that are out of bounds as well as bullets that have existed longer than their specified lifespan. The bullets are rendered, as well as some debug text.

The main update loop for any canvas runs *every 20 ms*, or *50 FPS*.

Now, Plural Modules. Plural Modules contain user-defined code. All Plural Modules *MUST* implement the `getPluralController(currgame, canvasid)` function, which returns a Plural Object to be associated with the provided canvas ID. As of the current version, the Plural Object's role is mostly ceremonial since only one attack at a time is currently supported (more may be supported in the future by giving Single objects a `finished` field and by having Plural objects check the status of the current Single). However, all Plural objecs must have a `step`, list of `singles`, `update()`, and `remove()`. The step starts at 0 and refers to which single is currently in use. The update function calls the current single to update, while remove clears the array of Single objects.

Single Objects are the meat of the system, controlling the happenings of a given Danmakanvas Instance. Each Single should have a list of `tasks`, though actually using this depends on whether or not such tasks are needed. As with the pattern, they must also have an `update()` and `remove()`, where update performs updates every 20 ms and handles task cleanup. 

Tasks are the smallest unit of control in Danmakanvas. Methods of creating subtasks are possible but require manual cleanup if done as the Single is only responsible for cleaning up tasks it creates itself. Tasks should have a `counter`, `maxcounter`, and `finished` state as well as the usual `update()` and `remove()`. `maxcounter` is a construct used for terminating the task after a given number of frames have passed. The counter refers to the frame count. The finished state should be set to true if the counter reaches the max counter - this allows the Single to properly dispose of the task on its update. All objects created by a task should be deleted in the tasks' `remove()` function.

### Programming using Danmakanvas

All programming is done via the Plural Modules. In a given Plural Module, ensure that `getPluralController()` has a case for the canvas ID you are targeting, and ensure that it receives a Plural object. The Template should assist in setting up the basics.

Note how `currgame` is passed along as a parameter for all Plurals, Singles, and Tasks. This is because bullets must be manually inserted into the current Danmakanvas Instance's list of bullets in order to function. Besides this, any number of arbitrary parameters may be used. Please avoid using `this.` for custom variables if possible, as they are not part of the object specifications.

Danmakanvas comes with a number of potentially useful functions built into the main game/Danmakanvas Instance object (`currgame`). Examples include `isinbounds()`, `toRadians()`, `contains()`, and `isEmpty()`, which are included for use in the main engine but may be used elsewhere. 

For actual control, `everyinterval()` allows for certain actions to be performed at a set rate. For example, `everyinterval(5)` will return true every five frames (e.g. 100, 200, 300, 400ms). Please note that Danmakanvas runs at 50 FPS rather than 60 FPS.

Now, for spawning bullets. We can run `new EnemyShot()` to accomplish this (or one of the provided bullet creation functions). The constructor contains a lot of parameters - `EnemyShot(x, y, speed, angle, accel, maxspeed, color, brad, srad, swid, hitbox, vanishtime, currgame)`. Most of these are obvious. Note that Danmakanvas currently only supports POSITIVE ACCELERATION. Slowing down a bullet to a lower max speed does NOT currently work. Color is a string, supporting all standard HTML supporte colors - including rgb(r,g,b) and hex notation (with the #). The next three parameters control the visuals of the bullet - the bullet radius, stroke radius, and stroke width. Hitbox is currently not used but is locked in for futureproofing. 

The vanishtime field of the bullet (EnemyShot) object behaves in a special manner. If the provided value is greater than 0, then that denotes how many frames the bullet will exist for before deleting. Otherwise the bullet will be deleted when it exits the game bounds. By default, the game bounds are a 32 pixel buffer around the canvas.

Like all JavaScript objects, it is possible to define a function and associate that function as part of the shot object. By default, the EnemyShot update runs a basic update that updates position, acceleration, and exist time. The EnemyShot `customupdate()` function can be manually overridden to induce interesting behavior.

Note that as of v2.2-dev3, there are a variety of bullet creation functions that can be used instead of the base constructor. These functions all automatically handle adding a bullet to the Danmakanvas Instance currently running. It is recommended that these be used instead of the default constructor in all cases.

### Available Functions

The following functions are provided as part of the Danmakanvas Instance (`currgame`, passed by reference as a parameter):

* `isinbounds(obj, wd, ht)` - Returns true if bullet is in the rectange created with (0, 0) and (wd, ht) with a 32 pixel buffer; false otherwise  
* `toRadians(i)` - Returns i converted from degrees to radians  
* `contains(a, obj)` - Returns true if a is present in the provided object; false otherwise  
* `isEmpty(obj)` - Returns true if the provided object is empty; false otherwise  
* `everyinterval(n)` - Returns true if the current frame count % n = 0; false otherwise  

The following functions are provided as Object Constructors:  
Note that these should not be used as they have proper wrappers that reduce some of the overhead. If they must be used, please note that they must be called with `new`.  

* `EnemyShot(x, y, speed, angle, accel, maxspeed, color, brad, srad, swid, hitbox, vanishtime, currgame)` - Creates an enemy shot with the provided parameters and returns the created object  
* `DMKText(x, y, fillStyle, fontsize, font, textAlign, content, currgame)` - Creates a Danmakanvas text object with the provided parameters and returns the created object.  

The following functions are provided as part of the Danmakanvas 'API':

General Functions:

* `GetCenterX(currgame)` - Returns half the width of the canvas  
* `GetCenterY(currgame)` - Returns half the height of the canvas  
* `DeleteShot(bullet, currgame)` - Deletes the bullet from the game  
* `GetDistanceObjObj(a, b)` - Returns the distance from object a to object b  
* `GetDistanceCoordObj(x1, y1, b)` - Returns the distance from (x1, y1) to object b  
* `GetDistanceCoordCoord(x1, y1, x2, y2)` - Returns the distance from (x1, y1) to (x2, y2)  
* `GetAngleObjObj(a, b)` - Returns the angle from object a to object b  
* `GetAngleCoordObj(x1, y1, b)` - Returns the angle from (x1, x2) to object b  
* `GetAngleCoordCoord(x1, y1, x2, y2)` - Returns the angle from (x1, x2) to object b  

Bullet Object Creation Functions:

- A1 variants set acceleration and max speed to 0 and set vanishtime to -1 (bullets only delete when out of bounds)
- All bullet creation functions add the bullets to the game's list of bullets automatically

* `CreateShotA1(n, x, y, speed, angle, color, brad, srad, swid, hitbox, currgame)` - Creates a bullet  
* `CreateShotA2(n, x, y, speed, angle, accel, maxspeed, color, brad, srad, swid, hitbox, vanishtime, currgame)` - Creates a bullet  
* `CreateRingA1(n, x, y, speed, angle, color, brad, srad, swid, hitbox, currgame)` - Creates a n-way ring  
* `CreateRingA2(n, x, y, speed, angle, accel, maxspeed, color, brad, srad, swid, hitbox, vanishtime, currgame)` - Creates a n-way ring  
* `CreateSpreadA1(n, angoffset, x, y, speed, angle, color, brad, srad, swid, hitbox, currgame)` - Creates an n-way spread with the provided angle offset  
* `CreateSpreadA2(n, angoffset, x, y, speed, angle, accel, maxspeed, color, brad, srad, swid, hitbox, vanishtime, currgame)` - Creates an n-way spread with the provided angle offset  
* `CreateStackA1(n, spdoffset, x, y, speed, angle, color, brad, srad, swid, hitbox, currgame)` - Creates an n-stack with the provided speed offset  
* `CreateStackA2(n, spdoffset, x, y, speed, angle, accel, maxspeed, color, brad, srad, swid, hitbox, vanishtime, currgame)` - Creates an n-stack with the provided speed offset  
* `CreateRingStackA1(n, m, spdoffset, x, y, speed, angle, color, brad, srad, swid, hitbox, currgame)` - Combination of CreateRingA1 and CreateStackA1  
* `CreateRingStackA2(n, m, spdoffset, x, y, speed, angle, accel, maxspeed, color, brad, srad, swid, hitbox, vanishtime, currgame)` - Combination of CreateRingA1 and CreateStackA1  
* `CreateSpreadStackA1(n, m, angoffset, spdoffset, x, y, speed, angle, color, brad, srad, swid, hitbox, currgame)` - Combination of CreateSpreadA1 and CreateStackA1  
* `CreateSpreadStackA2(n, m, angoffset, spdoffset, x, y, speed, angle, accel, maxspeed, color, brad, srad, swid, hitbox, vanishtime, currgame)` - Combination of CreateSpreadA1 and CreateStackA1  

Other Object Creation Functions:

* `CreateText(x, y, fillStyle, fontsize, font, textAlign, content, currgame)` - Creates a text object  

Object Functions:

* `SetShotGraphic(shot, graphic, color, brad, srad, srad2, swid, directed, rotation)` - Changes the graphics settings for an Enemy Shot  

### Bullet Graphics

Bullet Graphics default to a circle for the stroke. However, by using `SetShotGraphic()`, this can be changed.

Modes:  
* `CIRCLE` - Default. Stroke renders as a circle with `srad` radius  
* `DIAMOND` - Used for squares and rhombi. Stroke renders as a rhombus with `srad` controlling the principal axis in the direction it faces and `srad2` controlling the secondary axis perpendicular to the direction it faces. It is recommended to use the `directed` field for rhombi.  
* `TRIANGLE` - Used for isoceles and equilateral triangles. Stroke renders as a triangle with `srad` controlling the principal axis in the direction it faces and `srad2` controlling the distance to the other two points. It is recommended to use the `directed` field for triangles.  
* `OVAL` - Used for ellipses. Stroke renders as an ellipse with `srad` controlling the principal axis for the direction it faces and `srad2` controlling the secondary axis perpendicular to the direction it faces. It is recommended to use the `directed` field for ellipses.  
* `ARROWHEAD` - Used for arrowhead bullets. Stroke renders as an arrowhead with `srad` controlling the principal axis for the direction it faces and `srad2` controlling the secondary axis perpendicular to the direction it faces. The stroke is moved 1/3 of `srad` towards the bullet center. It is recommended to use the `directed` field for ellipses.  

### Miscellaneous

* There is no background color set for the canvas. Please use CSS for colors. We recommend a black or near-black background for the canvas as text defaults to white and as bullets tend to look better on a black background.

### Examples

Example scripts are documented here. In recommended reading order, these are:

1. [Template](https://sparen.github.io/Danmakanvas/Examples/Template.html) - Basic template for all Danmakanvas scripts  
2. [Example_Interval](https://sparen.github.io/Danmakanvas/Examples/Example_Interval.html) - Example showcasing `everyinterval()` usage as well as basic bullet spawning  
3. [Example_Override](https://sparen.github.io/Danmakanvas/Examples/Example_Override.html) - Example showcasing overriding of EnemyShot `customupdate()` to induce special behavior  
4. [Example_RingSpreadStack](https://sparen.github.io/Danmakanvas/Examples/Example_RingSpreadStack.html) - Example showcasing usage of GetCenterX/Y and some ring and stack functions. Also covers `customupdate()` overriding with these functions and provides more detail  
5. [Example_Graphics](https://sparen.github.io/Danmakanvas/Examples/Example_Graphics.html) - Example showcasing usage of `SetShotGraphics()` with rhombi and squares. Demonstrates both the `directed` option and the standard rotating option.  
6. [Example_Aimed](https://sparen.github.io/Danmakanvas/Examples/Example_Aimed.html) - Example showcasing usage of bullet aiming and distance functions.  

### Future Work

Listed in no particular order:

- Lasers
- Support for multiple singles with progression  



Changelog.md
# Changelog

Only versions past 2.1 have their changes documented here.

Source code and usage examples can be found on Github.

### v2.3-dev1 [Mar 25, 2020]
* Added `stopGame()` functionality to end execution of a canvas  
* Trying to create a new game now destroys the old one and creates a new one if one already exists, unlike before where nothing happened  
* Added `clearFrame` functionality to `stopGame()`  

### v2.2-dev7 [Nov 15, 2019]
* Added Distance and Angle functions  

### v2.2-dev6 [Nov 12, 2019]
* Added support for Oval and Arrowhead shapes  
* Enabled ability to disable clearing the playing field  

### v2.2-dev5 [Nov 11, 2019]
* Added support for Triangle shape  
* Decoupled font size from font in Text  

### v2.2-dev4 [Nov 08, 2019]
* Added `DMKText` object and general use `CreateText()` function for text object creation  
* EnemyShot objects now have fields for their render fields (brad, srad, swid) as well as new graphic and swid2 parameters for custom graphics  
* Implemented `SetShotGraphic()` and added support for the Rhombus/Diamond shape  

### v2.2-dev3 [Nov 03, 2019]
* Moved existing A1 variants to the A2 designation and created separate A1 designations for all existing bullet creation functions  
* Implemented `CreateShotA1()` and `CreateShotA2()`  
* Implemented `DeleteShot()`  

### v2.2-dev2 [Nov 02, 2019]

* Implemented `CreateRingA1()`, `CreateSpreadA1()`, `CreateStackA1()`, `CreateRingStackA1()`, and `CreateSpreadStackA1()`  
* Added `GetCenterX()` and `GetCenterY()`  

### v2.2-dev1 [Nov 01, 2019]

* Repository started  
* `var` -> `let` in new templates for Plural usage  
* Refactoring of main engine components to reduce naming conflicts  
* Fixed frameNo not being reset on canvas restart  
* Added ability to override an EnemyShot's update, allowing for custom behavior  

### v2.1.3 [Sep 11, 2018]

* Bullet Count text now uses canvas height instead of hardcoding.  

### v2.1.2 [Sep 10, 2018]

* isinbounds now uses canvas dimensions instead of hardcoding.  
* Version number now stored at top of file as variable for easy access and reference.  

### v2.1.1 [May 03, 2018]

* Fixed issue where clicking 'Run Danmakanvas Simulation' multiple times would start new instances of the simulation on the same canvas. Issue was a porting issue from 2.0, where startedplurals was not global and was accidentally made local to a game instance (where it was effectively useless). New clicks no longer boot up a new instance. [Thanks to Arcvasti]  
* Note that the 2.0 behavior of reset-on-click is not implemented in 2.1.1 and will require more changes in order to work. Whether or not this feature is necessary is a different issue entirely.  

### v2.1 [Apr 29, 2018]

* New version of Danmakanvas that has each game as a new instance, allowing for multiple canvases on the same page.  
* Plural/Single format is different and old scripts for 2.0 are not compatible with 2.1, as information on the current game must be threaded into the plurals and singles.  


Examples/PluralModule_Template.js
//Plural Module Template

"use strict";

// Controller that determines which attacks to display
function getPluralController(currgame, canvasid) {
    switch(canvasid) {
        case "gamecanvas_1":
            return new Plural_1(currgame);
            break;
        default:
            console.log("getPluralController(): Canvas ID " + canvasid + " could not be recognized. Please check to make sure that the canvas ID is correct and/or supported.");
    }
}

// Constructor for a Plural
function Plural_1(currgame) {
    this.step = 0; // Default to first single in array
    this.singles = [new Single_1(currgame)];
    this.update = function () {
        this.singles[this.step].update();
    }
    this.remove = function () {
        this.singles = [];
    }
}

// Constructor for a Single
function Single_1(currgame) {
    this.tasks = [];
    // Push Main Tasks here:
    this.tasks.push(new Single_1_Task_Template(currgame));
    this.update = function () { // Main Loop for a given Danmakanvas Instance
        // In update, push tasks that run every x frames
        // Remove completed tasks
        let tasktoremove = [];
        let i;
        for (i = 0; i < this.tasks.length; i += 1) {
            this.tasks[i].update();
            if (this.tasks[i].finished) {
                tasktoremove.push(i);
                this.tasks[i].remove();
            }
        }
        for (i = tasktoremove.length - 1; i >= 0; i -= 1) {
            this.tasks.splice(tasktoremove[i], 1);
        }
    }
    this.remove = function () {
        this.tasks = [];
    }
}

function Single_1_Task_Template(currgame) {
    this.counter = 0;
    this.maxcounter = -1; // Maximum time allowed for task to run. Use -1 for non-terminating tasks
    this.finished = false;
    this.update = function () {
        this.counter += 1;
        if (this.counter === this.maxcounter) {
            this.finished = true;
        }
    }
    this.remove = function () { // Deconstructor. Called by the Single object. Destroy any subtasks or objects created by the task here.
    }
}

Examples/PluralModule_Example_Aimed.js
//Plural Module - Example - Aimed

"use strict";

// Controller that determines which attacks to display
function getPluralController(currgame, canvasid) {
    switch(canvasid) {
        case "gamecanvas_1":
            return new Plural_1(currgame);
            break;
        default:
            console.log("getPluralController(): Canvas ID " + canvasid + " could not be recognized. Please check to make sure that the canvas ID is correct and/or supported.");
    }
}

// Constructor for a Plural
function Plural_1(currgame) {
    this.step = 0; // Default to first single in array
    this.singles = [new Single_1(currgame)];
    this.update = function () {
        this.singles[this.step].update();
    }
    this.remove = function () {
        this.singles = [];
    }
}

// Constructor for a Single
function Single_1(currgame) {
    this.tasks = [];
    // Push Main Tasks here:
    let angleT = 0;
    let mainshot = CreateShotA1(GetCenterX(currgame) + 160*Math.cos(angleT), GetCenterY(currgame) + 120*Math.sin(angleT), 0, 0, "#FF0000", 8, 12, 2, 0, currgame);
    let disttext = CreateText(4, 40, "white", "12px", "Arial", "left", "", currgame);
    this.update = function () { // Main Loop for a given Danmakanvas Instance
        mainshot.x = GetCenterX(currgame) + 160*Math.cos(angleT);
        mainshot.y = GetCenterY(currgame) + 120*Math.sin(angleT);
        disttext.content = "Distance: " + GetDistanceCoordObj(GetCenterX(currgame), GetCenterY(currgame), mainshot);
        if (currgame.everyinterval(10)) {
            let currangle = GetAngleCoordObj(GetCenterX(currgame), GetCenterY(currgame), mainshot);
            CreateRingStackA1(8, 5, 0.2, GetCenterX(currgame), GetCenterY(currgame), 6, currangle, "#00FFFF", 4, 6, 1, 0, currgame);
        }
        angleT += 0.02;
        // Remove completed tasks
        let tasktoremove = [];
        let i;
        for (i = 0; i < this.tasks.length; i += 1) {
            this.tasks[i].update();
            if (this.tasks[i].finished) {
                tasktoremove.push(i);
                this.tasks[i].remove();
            }
        }
        for (i = tasktoremove.length - 1; i >= 0; i -= 1) {
            this.tasks.splice(tasktoremove[i], 1);
        }
    }
    this.remove = function () {
        this.tasks = [];
    }
}

// Function telling the target to increment its angle by wvel every frame. WARNING: Angular velocity is in radians!
function applywvel(target, wvel) {
    target.angle += wvel;
}


Examples/PluralModule_Example_Graphics.js
//Plural Module - Example - Graphics

"use strict";

// Controller that determines which attacks to display
function getPluralController(currgame, canvasid) {
    switch(canvasid) {
        case "gamecanvas_1":
            return new Plural_1(currgame);
            break;
        default:
            console.log("getPluralController(): Canvas ID " + canvasid + " could not be recognized. Please check to make sure that the canvas ID is correct and/or supported.");
    }
}

// Constructor for a Plural
function Plural_1(currgame) {
    this.step = 0; // Default to first single in array
    this.singles = [new Single_1(currgame)];
    this.update = function () {
        this.singles[this.step].update();
    }
    this.remove = function () {
        this.singles = [];
    }
}

// Constructor for a Single
function Single_1(currgame) {
    this.tasks = [];
    // Push Main Tasks here:
    let currdir = 1;
    this.update = function () { // Main Loop for a given Danmakanvas Instance
        if (currgame.everyinterval(50)) { 
            let randangle = Math.random() * Math.PI*2;
            let bullets = CreateRingStackA2(20, 3, 0.2, GetCenterX(currgame), GetCenterY(currgame), 0, randangle, 0.01, 12, "#FF00FF", 2, 3, 1, 1, -1, currgame);
            let objctr = 0;
            for (objctr = 0; objctr < bullets.length; objctr += 1) {
                let currbullet = bullets[objctr];
                SetShotGraphic(currbullet, "DIAMOND", "#FF00FF", 1, 8, 4, 1, true, 0);
                currbullet.customupdate = function() {applywvel(currbullet, 0.05 * currdir);}
            }
            let bullets2 = CreateRingStackA2(20, 2, 0.2, GetCenterX(currgame), GetCenterY(currgame), 0.1, randangle, 0.01, 12, "#00FFFF", 2, 3, 1, 1, -1, currgame);
            let objctr2 = 0;
            for (objctr2 = 0; objctr2 < bullets2.length; objctr2 += 1) {
                let currbullet = bullets2[objctr2];
                SetShotGraphic(currbullet, "DIAMOND", "#00FFFF", 1, 6, 6, 1, true, 0);
                currbullet.customupdate = function() {applywvel(currbullet, 0.05 * currdir);}
            }
            let bullets3 = CreateRingStackA2(3, 12, 0.1, GetCenterX(currgame), GetCenterY(currgame), 2, randangle, 0.01, 12, "#FFFF00", 2, 3, 1, 1, -1, currgame);
            let objctr3 = 0;
            for (objctr3 = 0; objctr3 < bullets3.length; objctr3 += 1) {
                let currbullet = bullets3[objctr3];
                SetShotGraphic(currbullet, "DIAMOND", "#FFFF00", 1, 6, 6, 1, false, 0.05);
                let angvel = 0.015 * currdir;
                currbullet.customupdate = function() {applywvel(currbullet, angvel);}
            }
            let bullets4 = CreateRingStackA2(3, 12, 0.1, GetCenterX(currgame), GetCenterY(currgame), 2, randangle, 0.01, 12, "#FF8800", 2, 3, 1, 1, -1, currgame);
            let objctr4 = 0;
            for (objctr4 = 0; objctr4 < bullets4.length; objctr4 += 1) {
                let currbullet = bullets4[objctr4];
                SetShotGraphic(currbullet, "DIAMOND", "#FF8800", 1, 6, 6, 1, false, -0.05);
                let angvel = -0.015 * currdir;
                currbullet.customupdate = function() {applywvel(currbullet, angvel);}
            }
            currdir *= -1; // Flip angular velocity direction each time
        }
        // Remove completed tasks
        let tasktoremove = [];
        let i;
        for (i = 0; i < this.tasks.length; i += 1) {
            this.tasks[i].update();
            if (this.tasks[i].finished) {
                tasktoremove.push(i);
                this.tasks[i].remove();
            }
        }
        for (i = tasktoremove.length - 1; i >= 0; i -= 1) {
            this.tasks.splice(tasktoremove[i], 1);
        }
    }
    this.remove = function () {
        this.tasks = [];
    }
}

// Function telling the target to increment its angle by wvel every frame. WARNING: Angular velocity is in radians!
function applywvel(target, wvel) {
    target.angle += wvel;
}


Examples/PluralModule_Example_Interval.js
//Plural Module - Example - Interval

"use strict";

// Controller that determines which attacks to display
function getPluralController(currgame, canvasid) {
    switch(canvasid) {
        case "gamecanvas_1":
            return new Plural_1(currgame);
            break;
        default:
            console.log("getPluralController(): Canvas ID " + canvasid + " could not be recognized. Please check to make sure that the canvas ID is correct and/or supported.");
    }
}

// Constructor for a Plural
function Plural_1(currgame) {
    this.step = 0; // Default to first single in array
    this.singles = [new Single_1(currgame)];
    this.update = function () {
        this.singles[this.step].update();
    }
    this.remove = function () {
        this.singles = [];
    }
}

// Constructor for a Single
function Single_1(currgame) {
    this.tasks = [];
    // Push Main Tasks here:
    let j = 1;
    for (j = 1; j < 50; j += 1) {
        this.tasks.push(new Single_1_Task_Interval(currgame, j));
    }
    this.update = function () { // Main Loop for a given Danmakanvas Instance
        // In update, push tasks that run every x frames
        // Remove completed tasks
        let tasktoremove = [];
        let i;
        for (i = 0; i < this.tasks.length; i += 1) {
            this.tasks[i].update();
            if (this.tasks[i].finished) {
                tasktoremove.push(i);
                this.tasks[i].remove();
            }
        }
        for (i = tasktoremove.length - 1; i >= 0; i -= 1) {
            this.tasks.splice(tasktoremove[i], 1);
        }
    }
    this.remove = function () {
        this.tasks = [];
    }
}

function Single_1_Task_Interval(currgame, interval) {
    this.counter = 0;
    this.maxcounter = -1; // Maximum time allowed for task to run. Use -1 for non-terminating tasks
    this.finished = false;
    this.update = function () {
        if (currgame.everyinterval(interval)) { 
            var newshot = new CreateShotA2(640/51/2 + 640/51*interval, 32, 3, Math.PI/2, 0, 5, "rgb(" + (Math.sin(interval)*255) + "," + (Math.cos(interval)*255) + "," + (Math.sin(interval*2)*255) + ")", 3, 5, 0.75, 4, -1, currgame);
        }

        this.counter += 1;
        if (this.counter === this.maxcounter) {
            this.finished = true;
        }
    }
    this.remove = function () { // Deconstructor. Called by the Single object. Destroy any subtasks or objects created by the task here.
    }
}

Examples/PluralModule_Example_Override.js
//Plural Module - Example - Override

"use strict";

// Controller that determines which attacks to display
function getPluralController(currgame, canvasid) {
    switch(canvasid) {
        case "gamecanvas_1":
            return new Plural_1(currgame);
            break;
        default:
            console.log("getPluralController(): Canvas ID " + canvasid + " could not be recognized. Please check to make sure that the canvas ID is correct and/or supported.");
    }
}

// Constructor for a Plural
function Plural_1(currgame) {
    this.step = 0; // Default to first single in array
    this.singles = [new Single_1(currgame)];
    this.update = function () {
        this.singles[this.step].update();
    }
    this.remove = function () {
        this.singles = [];
    }
}

// Constructor for a Single
function Single_1(currgame) {
    this.tasks = [];
    // Push Main Tasks here:
    this.tasks.push(new Single_1_Task_AngularVelocity(currgame));
    this.update = function () { // Main Loop for a given Danmakanvas Instance
        // In update, push tasks that run every x frames
        if (currgame.everyinterval(50)) { 
            this.tasks.push(new Single_1_Task_Arrows(currgame));
        }
        // Remove completed tasks
        let tasktoremove = [];
        let i;
        for (i = 0; i < this.tasks.length; i += 1) {
            this.tasks[i].update();
            if (this.tasks[i].finished) {
                tasktoremove.push(i);
                this.tasks[i].remove();
            }
        }
        for (i = tasktoremove.length - 1; i >= 0; i -= 1) {
            this.tasks.splice(tasktoremove[i], 1);
        }
    }
    this.remove = function () {
        this.tasks = [];
    }
}

// Function telling the target to increment its angle by wvel every frame. WARNING: Angular velocity is in radians!
function applywvel(target, wvel) {
    target.angle += wvel;
}

function Single_1_Task_AngularVelocity(currgame) {
    this.counter = 0;
    this.maxcounter = -1; // Maximum time allowed for task to run. Use -1 for non-terminating tasks
    this.finished = false;
    let attackcounter = 0;
    this.update = function () {
        if (currgame.everyinterval(4)) { 
            let i = 0;
            if (attackcounter % 2 == 0) {
                for (i = 0; i < 30; i += 1) {
                    let newshot = new CreateShotA2(320, 240, 3, Math.PI*2/30*i, 0, 5, "#00FFFF", 2, 4, 0.5, 4, -1, currgame);
                    newshot.customupdate = function() {applywvel(newshot, 0.01);}
                }
                attackcounter += 1;
            } else {
                for (i = 0; i < 30; i += 1) {
                    let newshot = new CreateShotA2(320, 240, 3, Math.PI*2/30*i, 0, 5, "#FF00FF", 2, 4, 0.5, 4, -1, currgame);
                    newshot.customupdate = function() {applywvel(newshot, -0.01);}
                }
                attackcounter += 1;
            }
        }

        this.counter += 1;
        if (this.counter === this.maxcounter) {
            this.finished = true;
        }
    }
    this.remove = function () { // Deconstructor. Called by the Single object. Destroy any subtasks or objects created by the task here.
    }
}

// Function telling the target to zigzag by dt every f frames
function applyzigzag(target, f, dt) {
    if (target.existtime % (2*f) == 0) {
        target.angle += dt;
    }
    if (target.existtime % (2*f) == f) {
        target.angle -= dt;
    }
}

function Single_1_Task_Arrows(currgame) {
    this.counter = 0;
    this.maxcounter = 25; // Maximum time allowed for task to run. Use -1 for non-terminating tasks
    this.finished = false;
    let initangle = Math.random() * Math.PI*2;
    this.update = function () {
        let i = 0;
        for (i = 0; i < 8; i += 1) {
            let newshot = new CreateShotA2(320, 240, 3, initangle + Math.PI*2/8*i, 0, 5, "#FFFF00", 2, 4, 0.5, 4, -1, currgame);
            newshot.customupdate = function() {applyzigzag(newshot, 15, Math.PI/3);}
        }

        this.counter += 1;
        if (this.counter === this.maxcounter) {
            this.finished = true;
        }
    }
    this.remove = function () { // Deconstructor. Called by the Single object. Destroy any subtasks or objects created by the task here.
    }
}